src/coprocessor/dag/expr/builtin_encryption.rs

```diff
@@ -91,13 +95,93 @@ impl ScalarFunc {
         };
         Ok(Some(Cow::Owned(sha2)))
     }
+
+    #[inline]
+    pub fn compress<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &[Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let input = try_opt!(self.children[0].eval_string(ctx, row));
+        if input.is_empty() {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+        let mut e = ZlibEncoder::new(input.as_ref(), Compression::default());
+        let mut vec = Vec::with_capacity(1024);
+        match e.read_to_end(&mut vec) {
+            Ok(len) => {
+                let mut cap = 4 + len;
+                // according to MySQL doc: If the string ends with space,
+                // an extra . character is added to avoid problems with endspace trimming
+                if vec[len - 1] == 32 {
+                    vec.push(b'.');
+                    cap += 1;
+                }
+
+                let mut wtr = Vec::with_capacity(cap);
```

Comment:
OMG GitHub lost my comment..

There is several re-allocation here, maybe it can be eliminated in the following ways:

1. Create a Vec with capacity=original length (we asserted that in most cases compressed length < original length)
2. Resize to 4 bytes (fill with 0)
3. encoder = ZlibEncoder::new(vec)
4. encoder.write_all(..)
5. encoder.finish()
6. append `.` if necessary
7. modify first 4 bytes

Reply:
@breeswish thanks. code updated. PTAL